### PR TITLE
Development

### PR DESCRIPTION
**Pull request by Ash-Manoj**
Branch with Whitbox GIS python tools for using with Automatic CATFLOW delineation tool.
Possible question regarding datatype - 'None' in python. (In aspect tool/ zfactor)
Should the .yml definition accept it as string or a default datatype None is required?

### CODE REVIEW

File: src/run.py [hunk 1]
```diff
@@ -31,7 +31,118 @@
     print(f"Filling depressions in DEM '{inp}'...",end='',flush=True)
     wblib.fill(inp,out,flats)
     print('done.')
+
+ # Aspect tool
+elif toolname == 'aspect':
+    # get the parameters
+    try:
+        inp = kwargs['filled_DEM']
```
Comment (mmaelicke) [inline src/run.py:39]: I would go for just DEM, as the tool does also work with non-filled DEMs

File: src/run.py [hunk 2]
```diff
@@ -31,7 +31,118 @@
...
+    # run the whitebox fill_depression algorithm
+    print(f"Calculating Slope Aspect in DEM '{inp}'...",end='',flush=True)
+    wblib.aspect(inp,out,zfactor)
+    print('done.')    
+
+ # Flow Accumulation tool
+elif toolname == 'flow_accumulation_d8':
+    # get the parameters
+    try:
+        inp = kwargs['filled_DEM']
```
Comment (mmaelicke) [inline src/run.py:55]: same here: DEM is technically enough. filled dem is just best practice

File: src/run.py [hunk 3]
```diff
@@ -31,7 +31,118 @@
...
+        out = '/out/flow_accu_DEM.tif'
```
Comment (mmaelicke) [inline src/run.py:56]: `flow_accumulation.tif`

File: src/run.py [hunk 4]
```diff
@@ -31,7 +31,118 @@
...
+    # run the whitebox  algorithm
+    print(f"Calculating Flow Accumulation  DEM '{inp}'...",end='',flush=True)
+    wblib.accu_d8(inp,out,type,log)
+    print('done.')       
     
+ # Flow Direction tool
+elif toolname == 'flow_direction_d8':
+    # get the parameters
+    try:
+        inp = kwargs['filled_DEM']
```
Comment (mmaelicke) [inline src/run.py:72]: `dem`

File: src/run.py [hunk 5]
```diff
@@ -31,7 +31,118 @@
...
+        out = '/out/flow_dir_DEM.tif'
```
Comment (mmaelicke) [inline src/run.py:73]: `flow_direction.tif`

File: src/run.py [hunk 6]
```diff
@@ -31,7 +31,118 @@
...
+    # run the whitebox algorithm
+    print(f"Hillslope Extraction from  '{inp}'...",end='',flush=True)
+    wblib.hillslope(inp,out,stream)
+    print('done.')    
+
+ # Elevation to River tool
+elif toolname == 'elevation_stream':
+    # get the parameters
+    try:
+        inp = kwargs['filled_DEM']
```
Comment (mmaelicke) [inline src/run.py:119]: See above

File: src/run.py [hunk 7]
```diff
@@ -31,7 +31,118 @@
...
+    print(f"Elevation from River '{stream}'...",end='',flush=True)
+    wblib.elevation(inp,out,stream)
+    print('done.')          
+
+ # Distance to River tool
+elif toolname == 'distance_stream':
+    # get the parameters
+    try:
+        inp = kwargs['filled_DEM']
+        out = '/out/dist2riv.tif'
```
Comment (mmaelicke) [inline src/run.py:136]: See above

File: src/run.py [hunk 8]
```diff
@@ -31,7 +31,118 @@
...
+        print(str(e))
+        sys.exit(1)
+
+    # run the whitebox algorithm
+    print(f"Hillslope Extraction from  '{inp}'...",end='',flush=True)
+    wblib.hillslope(inp,out,stream)
+    print('done.')    
+
+ # Elevation to River tool
+elif toolname == 'elevation_stream':
```
Comment (mmaelicke) [inline src/run.py:116]: elevation and distance should go into one tool.

**Review by mmaelicke — CHANGES_REQUESTED** [review]
Nice work! Thanks.

You should revert https://github.com/VForWaTer/tool_whiteboxgis/pull/9/commits/bbfc8b77569320f50a89c9e1cc0a1e88cc3851f6 and add the `out` folder to the `.gitignore`. We don't want to have output data in the repo.

**Comment by mmaelicke** [comment]
> Branch with Whitbox GIS python tools for using with Automatic CATFLOW delineation tool.
> Possible question regarding datatype - 'None' in python. (In aspect tool/ zfactor)
> Should the .yml definition accept it as string or a default datatype None is required?

You can either make it `optional: true` and then in Python: 
```python
zfactor = kwargs.get('zfactor, None)
```

or, ie. used in a enumeration set it as string:

```yaml
foobar:
  type: enum
  values: 
    - option 1
    - option 2
    - Not selected
``` 

then in Python: 

```python
foobar = kwargs['footer']
if foobar == 'Not selected':
  foobar = None

# or in one line
foobar = kwargs['foobar'] if kwargs['foobar'] != 'Not selected' else None
```

File: src/run.py (hunk 1 above)
Comment (Ash-Manoj) [reply src/run.py:39]: I agree!
Changed in revised commit.

File: in/DEM_W22.tif.aux.xml [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+  <PAMRasterBand band="1">
+    <Metadata>
+      <MDI key="STATISTICS_MAXIMUM">307.42749023438</MDI>
+      <MDI key="STATISTICS_MEAN">272.55642546599</MDI>
+      <MDI key="STATISTICS_MINIMUM">247.1036529541</MDI>
+      <MDI key="STATISTICS_STDDEV">13.298994922735</MDI>
+      <MDI key="STATISTICS_VALID_PERCENT">65.13</MDI>
+    </Metadata>
+  </PAMRasterBand>
+</PAMDataset>
```
Comment (mmaelicke) [inline in/DEM_W22.tif.aux.xml:11]: I guess this can also be removed, as it will be recreated on opening the geoTIFF, but that is not super important

**Review by mmaelicke — APPROVED** [review]
Great thanks!

**Pull request merged**